### DISCUSSION TITLE
Fallback for framework components that fail to hydrate or crash in the client (similar to ErrorBoundary or onerror)

### DISCUSSION THREAD

**Discussion by mb21** [withastro/roadmap#724] [Proposal]
# Summary

For any number of reasons, framework components can either

1. fail to hydrate, or
2. crash / throw an exception (when server-side-rendering, or as late as any point in time after hydration on the client)

Would be great to have either a `on:error` or `set:fallback` attribute on the framework component, which Astro would invoke and put into the page where the island was supposed to be.

I would be fine confining this to in-browser behaviour only, as we control the environment on the server much more tightly (and in the worst case we could render a 500 page). However if things crash on the client, currently you have just an empty hole missing without the user realizing what's going on.

# Background & Motivation

For 2., React and Solid have an `ErrorBoundary` component, so solving this second point is not so urgent, but perhaps it's easy to add when covering 1. as well?

An example of 1. is if I have a syntax error in my client-side JavaScript, then the whole file including the ErrorBoundary will fail to be executed. One example is when you use optional chaining (`.?`) and the user's browser doesn't support the syntax:

- https://github.com/withastro/astro/issues/3136
- https://github.com/withastro/astro/issues/4471
- https://github.com/withastro/astro/issues/6158
- https://github.com/withastro/astro/issues/6096

Potentially, `window.onerror` could be used already now, but this seems rather hacky and no way of knowing what else will bubble in there (browser plugins, ads, etc.)

# Goals

- Framework components should fail gracefully
- Provide a fallback for older browsers that cannot run the island

# Non-Goals

- Provide a fallback for browsers without JavaScript enabled (for that case, `<noscript>` already works).

# Example

Consider loading the `InteractiveCounter` framework component in an `.astro.` file.

## set:fallback

If we wish to handle only 1. (hydration errors), then this simple syntax might suffice:

```jsx
<InteractiveCounter
  set:fallback={<h2>Oops, this widget failed to load</h2>}
  />
```

## on:error

If we wish to handle 2. (later exceptions) as well, syntax could be something like this:

```jsx
<InteractiveCounter on:error={(err, reload) => (
  <>
    <h2>Oops, this widget crashed</h2>
    <p>Error was {err?.message}</p>
    Try to <button onClick={reload}>reload the widget</button>
  </>
  )}
  />
```

**Comment by bholmesdev** [comment]
Thanks for the proposal! Agreed error handling would be a nice feature. I believe we have an (undocumented) `slot="fallback"` to handle render failures for `client:only` components:

```astro
<MyComponent client:only="react">
<p slot="fallback">This browser may have JS disabled?</p>
</MyComponent>
```

I do prefer an explicit directive though. Paging @natemoo-re who built that slot feature

**Reply by mb21** [reply]
Thanks, that's great to know! I just tried it, but pity indeed that it only works with `client:only`...

**Reply by mb21** [reply]
Come to think of it, JS-disabled seems to actually be a bit of a different use-case, that you can solve with the `<noscript>` HTML element. What my original proposal adresses is when the browser understands JavaScript in principle, but it failed to load or crashed.

**Comment by lilnasy** [comment]
While this discussions was started specifically for client errors, I think a common solution for server renders is required as well.

This was recently brought up in issues by @darioguarascio: https://github.com/withastro/astro/issues/8790.

There, @ahmad-ali14 presented a solution that works in some cases: https://github.com/withastro/astro/issues/8790#issuecomment-1758000448. The approach works when you know which component throws the error, and it is in the frontmatter of an astro component. A response is returned to make the component render nothing, but it can be replaced with `null` or `undefined` to the same effect. Similarly, you can return a string to render a failure message in place of the component.

**Reply by mb21** [reply]
Thanks, yeah that works when the error happens on the server. But I'm really more concerned about in-browser failures here (because the server I control, so I can just handle the kind of things that still can throw, like network errors, but the whole component shouldn't crash). So yes, if it's easy to do the same thing on the server as well, go for it. If not, I'd add that as a non-goal. (Btw. I've updated the above RFC with some links to example client issues.)

**Comment by matthewp** [comment]
Something like this could probably be possible, although not trivially so. The example you provided wouldn't work because Astro doesn't have "slot props", but an actual named slot could potentially work. But it's pretty unclear to me how you might have access to error in that case.

That's the API difficulty, implementation would also be tricky. I think we'd have to include the error slot as a `<template>`. We already do this for unused slots so perhaps it could just hook into that mechanism and in that case it might not bee so bad.

**Reply by mb21** [reply]
Thanks for the reply! Not sure I understood everything. The `ErrorBoundary` components of [React](https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary) and [Solid](https://www.solidjs.com/tutorial/flow_error_boundary) seem to work just fine? Would such a wrapper component be easier to implement?

Or is the problem that in Astro templates, you cannot use JSX/HTML wherever you would have a JavaScript expression? So you cannot use it as a component prop value or inside a function?
